metaflow/plugins/cards/component_serializer.py

```diff
@@ -16,11 +23,368 @@ def get_card_class(card_type):
+            # as `self.components.layout_last_changed_on`, then the component array itself
+            # has been modified. So we should force a re-render of the card.
+            self._last_layout_change = self.components.layout_last_changed_on
+        else:
+            self._card_proc("refresh")
+
+    @property
+    def components(self):
+        return self._components
+
+    def _warning(self, message):
+        msg = "[@card WARNING] %s" % message
+        self._logger(msg, timestamp=False, bad=True)
+
+    def _get_latest_data(self, final=False, mode=None):
+        """
+        This function returns the data object that is passed down to :
+        - `MetaflowCard.render_runtime`
+        - `MetaflowCard.refresh`
+        - `MetaflowCard.reload_content_token`
+
+        The return value of this function contains all the necessary state information for Metaflow Cards to make decisions on the following:
+        1. What components are rendered
+        2. Should the card be reloaded on the UI
+        3. What data to pass down to the card.
+
+        Parameters
+        ----------
+        final : bool, optional
+            If True, it implies that the final "rendering" sequence is taking place (which involves calling a `render` and a `refresh` function.)
+            When final is set the `render_seq` is set to "final" so that the reload token in the card is set to final
+            and the card is not reloaded again on the user interface.
+        mode : str
+            This parameter is passed down to the object returned by this function. Can be one of `render_runtime` / `refresh` / `render`
+
+        Returns
+        -------
+        dict
+            A dictionary of the form :
+            ```python
+            {
+                "user": user_data, # any passed to `current.card.refresh` function
+                "components": component_dict, # all rendered REALTIME_UPDATABLE components
+                "render_seq": seq,
+                # `render_seq` is a counter that is incremented every time `render_runtime` is called.
+                # If a metaflow card has a RELOAD_POLICY_ALWAYS set then the reload token will be set to this value
+                # so that the card reload on the UI everytime `render_runtime` is called.
+                "component_update_ts": self.components.layout_last_changed_on,
+                # `component_update_ts` is the timestamp of the last time the component array was modified.
+                # `component_update_ts` can get used by the `reload_content_token` to make decisions on weather to
+                # reload the card on the UI when component array has changed.
+                "mode": mode,
+            }
+            ```
+        """
+        seq = "final" if final else self._render_seq
+        # Extract all the runtime-updatable components as a dictionary
+        component_dict = {}
```

Comment:
I wonder if we need to similarly break down data to components so each component can have their own data section?

Reply:
I don't quite catch that. Can you expand on that?

Reply:
well, the parts that are refreshable are at the component level (not the card level if I got this right). Yet, the data file is at the card level. It feels like there is a mismatch there but maybe I missed something.